File: 2022/day08.py
with open('data.txt', 'r') as f:
    grid = [[int(c) for c in ln.strip()] for ln in f]

nr = len(grid)
nc = len(grid[0])

def foo():
    return [[False] * nc for _ in range(nr)]

b0, b1, b2, b3 = (foo() for _ in range(4))

import math

for r in range(nr):
    hi = -math.inf
    for c in range(nc):
        g = grid[r][c]
        if g > hi:
            hi = g
            b0[r][c] = True

for r in range(nr):
    hi = -math.inf
    for c in range(nc-1, -1, -1):
        g = grid[r][c]
        if g > hi:
            hi = g
            b1[r][c] = True

for c in range(nc):
    hi = -math.inf
    for r in range(nr):
        g = grid[r][c]
        if g > hi:
            hi = g
            b2[r][c] = True

for c in range(nc):
    hi = -math.inf
    for r in range(nr-1, -1, -1):
        g = grid[r][c]
        if g > hi:
            hi = g
            b3[r][c] = True

p1 = 0
for r in range(nr):
    for c in range(nc):
        p1 += b0[r][c] or b1[r][c] or b2[r][c] or b3[r][c]

print(p1)

# part 2

def bar():
    return [[0] * nc for _ in range(nr)]

d0, d1, d2, d3 = (bar() for _ in range(4))

for r in range(nr):
    old = grid[r][0]
    grid[r][0] = math.inf  # hack
    st = [0]
    d0[r][0] = 0
    for c in range(1, nc):
        g = grid[r][c]
        while grid[r][st[-1]] < g:
            st.pop()
        d0[r][c] = abs(c - st[-1])
        st.append(c)
    grid[r][0] = old

for r in range(nr):
    old = grid[r][nc-1]
    grid[r][nc-1] = math.inf
    st = [nc-1]
    d1[r][nc-1] = 0
    for c in range(nc-2, -1, -1):
        g = grid[r][c]
        while grid[r][st[-1]] < g:
            st.pop()
        d1[r][c] = abs(c - st[-1])
        st.append(c)
    grid[r][nc-1] = old

for c in range(nc):
    old = grid[0][c]
    grid[0][c] = math.inf  # hack
    st = [0]
    d2[0][c] = 0
    for r in range(1, nr):
        g = grid[r][c]
        while grid[st[-1]][c] < g:
            st.pop()
        d2[r][c] = abs(r - st[-1])
        st.append(r)
    grid[0][c] = old

for c in range(nc):
    old = grid[nr-1][c]
    grid[nr-1][c] = math.inf  # hack
    st = [nr-1]
    d3[nr-1][c] = 0
    for r in range(nr-2, -1, -1):
        g = grid[r][c]
        while grid[st[-1]][c] < g:
            st.pop()
        d3[r][c] = abs(r - st[-1])
        st.append(r)
    grid[nr-1][c] = old

best = 0
for r in range(nr):
    for c in range(nc):
        best = max(best, d0[r][c] * d1[r][c] * d2[r][c] * d3[r][c])

print(best)

File: 2022/day08.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Cell = [2]int
type Inc = func(*[2]int)
type Check = func([2]int, *[][]byte) bool
type Compute = func(*[][]byte, Cell, Inc, Inc, Check, Check) *[][]int

func makeGrid(nr int, nc int) *[][]int {
	grid := make([][]int, nr)
	for i := range grid {
		grid[i] = make([]int, nc)
	}
	return &grid
}

func computeCanSee(
	grid *[][]byte, start Cell, outerInc Inc, innerInc Inc,
	outerCheck Check, innerCheck Check,
) *[][]int {
	canSee := makeGrid(len(*grid), len((*grid)[0]))
	for innerStart := start; outerCheck(innerStart, grid); outerInc(&innerStart) {
		var hi byte = 0
		for inner := innerStart; innerCheck(inner, grid); innerInc(&inner) {
			x := (*grid)[inner[0]][inner[1]]
			if x > hi {
				(*canSee)[inner[0]][inner[1]] = 1
				hi = x
			}
		}
	}
	return canSee
}

func goUp(ind *[2]int)    { ind[0]-- }
func goDown(ind *[2]int)  { ind[0]++ }
func goLeft(ind *[2]int)  { ind[1]-- }
func goRight(ind *[2]int) { ind[1]++ }

func insideLeft(ind [2]int, _ *[][]byte) bool     { return ind[1] >= 0 }
func insideRight(ind [2]int, grid *[][]byte) bool { return ind[1] < len((*grid)[0]) }
func insideAbove(ind [2]int, _ *[][]byte) bool    { return ind[0] >= 0 }
func insideBelow(ind [2]int, grid *[][]byte) bool { return ind[0] < len(*grid) }

func readGrid(filename string) *[][]byte {
	var grid [][]byte
	fin, _ := os.Open(filename)
	defer fin.Close()
	sc := bufio.NewScanner(bufio.NewReader(fin))
	for sc.Scan() {
		grid = append(grid, []byte(sc.Text()))
	}
	return &grid
}

func elemAt(grid *[][]byte, ind [2]int) byte {
	return (*grid)[ind[0]][ind[1]]
}

func computeNumberSeen(
	grid *[][]byte, start Cell, outerInc Inc, innerInc Inc,
	outerCheck Check, innerCheck Check,
) *[][]int {
	numSeen := makeGrid(len(*grid), len((*grid)[0]))
	for innerStart := start; outerCheck(innerStart, grid); outerInc(&innerStart) {
		// because cannot see any beyond edge
		(*numSeen)[innerStart[0]][innerStart[1]] = 0
		// because there's a tree there that will always be seen
		stack := [][2]int{innerStart}
		oneAfterInnerStart := innerStart
		innerInc(&oneAfterInnerStart)
		for inner := oneAfterInnerStart; innerCheck(inner, grid); innerInc(&inner) {
			g := elemAt(grid, inner)
			for 1 < len(stack) && g > elemAt(grid, stack[len(stack)-1]) {
				stack = stack[:len(stack)-1]
			}
			top := stack[len(stack)-1]
			// one of these diffs is zero
			(*numSeen)[inner[0]][inner[1]] = abs(inner[0]-top[0]) + abs(inner[1]-top[1])
			stack = append(stack, inner)
		}
	}
	return numSeen
}

type Update = func(int, int, int, int, int) int

func generalSolve(grid *[][]byte, compute Compute, update Update) int {
	nr := len(*grid)
	nc := len((*grid)[0])
	fromLeft := compute(grid, [2]int{0, 0}, goDown, goRight, insideBelow, insideRight)
	fromRight := compute(grid, [2]int{0, nc - 1}, goDown, goLeft, insideBelow, insideLeft)
	fromAbove := compute(grid, [2]int{0, 0}, goRight, goDown, insideRight, insideBelow)
	fromBelow := compute(grid, [2]int{nr - 1, 0}, goRight, goUp, insideRight, insideAbove)

	p := 0
	for r := 0; r < nr; r++ {
		for c := 0; c < nc; c++ {
			p = update(p, (*fromLeft)[r][c], (*fromRight)[r][c], (*fromAbove)[r][c], (*fromBelow)[r][c])
		}
	}
	return p
}

func highScore(p int, a int, b int, c int, d int) int {
	score := a * b * c * d
	if score > p {
		return score
	} else {
		return p
	}
}

func part2(grid *[][]byte) int {
	return generalSolve(grid, computeNumberSeen, highScore)
}

func incrementIfAny(p int, a int, b int, c int, d int) int {
	if a == 1 || b == 1 || c == 1 || d == 1 {
		return p + 1
	} else {
		return p
	}
}

func part1(grid *[][]byte) int {
	return generalSolve(grid, computeCanSee, incrementIfAny)
}

func main() {
	grid := readGrid("/home/xdavidliu/Documents/aoc/input08.txt")
	fmt.Println("part 1 =", part1(grid)) // 1681
	fmt.Println("part 2 =", part2(grid)) // 201684
}
